Source/Core/Core/HW/SystemTimers.cpp

```diff
@@ -170,57 +248,29 @@ s64 GetLocalTimeRTCOffset()
   return s_localtime_rtc_offset;
 }
 
-static void PatchEngineCallback(u64 userdata, s64 cycles_late)
+double GetEstimatedEmulationPerformance()
 {
-  // We have 2 periods, a 1000 cycle error period and the VI period.
-  // We have to carefully combine these together so that we stay on the VI period without drifting.
-  u32 vi_interval = VideoInterface::GetTicksPerField();
-  s64 cycles_pruned = (userdata + cycles_late) % vi_interval;
-  s64 next_schedule = 0;
-
-  // Try to patch mem and run the Action Replay
-  if (PatchEngine::ApplyFramePatches())
+  u64 ts_now, ts_before;  // In microseconds
   {
-    next_schedule = vi_interval - cycles_pruned;
-    cycles_pruned = 0;
-  }
-  else
-  {
-    // The patch failed, usually because the CPU is in an inappropriate state (interrupt handler).
-    // We'll try again after 1000 cycles.
-    next_schedule = 1000;
-    cycles_pruned += next_schedule;
-  }
+    std::lock_guard<std::mutex> lk(s_emu_to_real_time_mutex);
+    size_t index_now = s_emu_to_real_time_index == 0 ? s_emu_to_real_time_ring_buffer.size() :
+                                                       s_emu_to_real_time_index - 1;
+    size_t index_before = s_emu_to_real_time_index;
 
-  CoreTiming::ScheduleEvent(next_schedule, et_PatchEngine, cycles_pruned);
-}
-
-static void ThrottleCallback(u64 last_time, s64 cyclesLate)
-{
-  // Allow the GPU thread to sleep. Setting this flag here limits the wakeups to 1 kHz.
-  Fifo::GpuMaySleep();
-
-  u32 time = Common::Timer::GetTimeMs();
+    ts_now = s_emu_to_real_time_ring_buffer[index_now];
+    ts_before = s_emu_to_real_time_ring_buffer[index_before];
+  }
 
-  int diff = (u32)last_time - time;
-  const SConfig& config = SConfig::GetInstance();
-  bool frame_limiter = config.m_EmulationSpeed > 0.0f && !Core::GetIsThrottlerTempDisabled();
-  u32 next_event = GetTicksPerSecond() / 1000;
-  if (frame_limiter)
+  if (ts_before == 0)
   {
-    if (config.m_EmulationSpeed != 1.0f)
-      next_event = u32(next_event * config.m_EmulationSpeed);
-    const int max_fallback = config.iTimingVariance;
-    if (abs(diff) > max_fallback)
-    {
-      DEBUG_LOG(COMMON, "system too %s, %d ms skipped", diff < 0 ? "slow" : "fast",
-                abs(diff) - max_fallback);
-      last_time = time - max_fallback;
-    }
-    else if (diff > 0)
-      Common::SleepCurrentThread(diff);
+    // Not enough data yet to estimate. We could technically provide an estimate based on a shorter
+    // time horizon, but it's not really worth it.
+    return 1.0;
   }
-  CoreTiming::ScheduleEvent(next_event - cyclesLate, et_Throttle, last_time + 1);
+
+  u64 delta_us = ts_now - ts_before;
+  double emulated_us = s_emu_to_real_time_ring_buffer.size() * 1000.0;  // For each emulated ms.
```

Comment:
This could actually be `constexpr`, right? Since the size of the array is known in advance.

Reply:
The compiler should optimize that out -- I'm not going to manually hint it, since this is not performance sensitive in any way.